File: golang/array_string/jumpGameII.go
package main

func jump(nums []int) int {
	if len(nums) <= 1 {
		return 0
	}

	jumps := 0      // Count of jumps needed
	currentEnd := 0 // End of the current jump range
	farthest := 0   // Farthest point we can reach

	for i := 0; i < len(nums)-1; i++ {
		// Update the farthest we can reach from the current position
		farthest = max(farthest, i+nums[i])

		// If we have reached the end of the current jump range
		if i == currentEnd {
			jumps++
			currentEnd = farthest // Update the end of the next jump range

			// Early exit if we can already reach the last index
			if currentEnd >= len(nums)-1 {
				break
			}
		}
	}

	return jumps
}

func max(a, b int) int {
	if a > b {
		return a
	}

	return b
}

File: python/array_string/jumpGameII.py
from typing import List


class Solution:
    def jump(self, nums: List[int]) -> int:
        if len(nums) <= 1:
            return 0

        jumps = 0
        current_end = 0
        farhest = 0

        for i in range(len(nums) - 1):
            farhest = max(farhest, i + nums[i])

            if i == current_end:
                jumps += 1

                current_end = farhest

                if current_end >= len(nums) - 1:
                    break

        return jumps
